core/src/main/java/io/questdb/cairo/CairoMetadata.java

```
@@ -0,0 +1,613 @@
+                sink.put('\t');
+                table.toSink(sink);
+                sink.put('\n');
+            }
+            close();
+            String s = sink.toString();
+            sink.clear();
+            tlSink.remove();
+            return s;
+        }
+    }
+
+    private class CairoMetadataWriter extends CairoMetadataReader implements Closeable, CairoMetadataRW, AutoCloseable {
+
+        public void clear() {
+            tables.clear();
+        }
+
+        @Override
+        public void close() {
+            lock.writeLock().unlock();
+        }
+
+        public void dropTable(@NotNull CharSequence tableName) {
+            tables.remove(tableName);
+            LOG.info().$("dropped metadata [table=").$(tableName).I$();
+        }
+
+        public void dropTable(@NotNull TableToken tableToken) {
+            dropTable(tableToken.getTableName());
+        }
+
+        @Override
+        public CairoTable getTable(@NotNull TableToken tableToken) {
+            CairoTable table = super.getTable(tableToken);
+            if (table == null) {
+                // hydrate cache
+                hydrateTable(tableToken, true);
+            }
+            table = super.getTable(tableToken);
+            return table;
+        }
+
+        /**
+         * This is dangerous and may clobber any concurrent metadata changes. This should only be used in last-resort cases,
+         * for example, for testing purposes.
+         */
+        @TestOnly
+        public void hydrateAllTables() {
+            ObjHashSet<TableToken> tableTokensSet = new ObjHashSet<>();
+            engine.getTableTokens(tableTokensSet, false);
+            ObjList<TableToken> tableTokens = tableTokensSet.getList();
+
+            if (tableTokens.size() == 0) {
+                LOG.error().$("could not hydrate metadata, there are no table tokens").$();
+                return;
+            }
+
+            TableToken tableToken = tableTokens.getQuick(0);
```

error handling is questionable; since this is used in restore from backup - instead of logging an error this method should hard fail and subsequently abort the restore